eth/watchers/pricefeedwatcher_test.go

```diff
@@ -0,0 +1,319 @@
+package watchers
+
+import (
+	"context"
+	"errors"
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/livepeer/go-livepeer/eth"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+type mockPriceFeedEthClient struct {
+	mock.Mock
+}
+
+func (m *mockPriceFeedEthClient) FetchPriceData() (eth.PriceData, error) {
+	args := m.Called()
+	return args.Get(0).(eth.PriceData), args.Error(1)
+}
+
+func (m *mockPriceFeedEthClient) Description() (string, error) {
+	args := m.Called()
+	return args.String(0), args.Error(1)
+}
+
+func TestPriceFeedWatcher_UpdatePrice(t *testing.T) {
+	priceFeedMock := new(mockPriceFeedEthClient)
+	defer priceFeedMock.AssertExpectations(t)
+
+	priceData := eth.PriceData{
+		RoundID:   10,
+		Price:     big.NewRat(3, 2),
+		UpdatedAt: time.Now(),
+	}
+	priceFeedMock.On("FetchPriceData").Return(priceData, nil).Once()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	w := &PriceFeedWatcher{
+		ctx:           ctx,
+		priceFeed:     priceFeedMock,
+		currencyBase:  "ETH",
+		currencyQuote: "USD",
+		priceUpdated:  make(chan eth.PriceData, 1),
+	}
+
+	require.NoError(t, w.updatePrice())
+	require.Equal(t, priceData, w.current)
+	select {
+	case updatedPrice := <-w.priceUpdated:
+		require.Equal(t, priceData, updatedPrice)
+	case <-time.After(2 * time.Second):
+		t.Error("Updated price hasn't been received on channel")
+	}
+}
+
+func TestPriceFeedWatcher_Watch(t *testing.T) {
```

Comment:
I wonder if we should do so many tests which are time-based + fake tickers. The drawback of them is that the tests can be flaky. Plus, with this fake ticker, I guess you had to add some complexity to the prod logic to make it testable....

Reply:
Yeah these were quite not the breeze to implement haha

> The drawback of them is that the tests can be flaky. 

The fake tickers make sure they are not flaky (nor take too long)!

> Plus, with this fake ticker, I guess you had to add some complexity to the prod logic to make it testable....

Yeah, you can check the changes I had to made in the prod code here:
- [for testing truncated ticker](https://github.com/livepeer/go-livepeer/pull/2972/commits/4db815345c4c0d45d8f671dc9d6efabd3b391c11#diff-91f8f3d3888f4535a8a9b736fbb8b471b29155db3611c1c520b5105d4e6a7dd3L137)
- [for testing `watch()`](https://github.com/livepeer/go-livepeer/pull/2972/commits/59450bf156a3a1a3234c426c2d3251c837ab9599#diff-91f8f3d3888f4535a8a9b736fbb8b471b29155db3611c1c520b5105d4e6a7dd3L98)

In the end I don't think it was too bad, but let me know if you have a better idea (like dropping the tests, I'd be happy to lol).